File: src/java/pulse/cdm/patient/actions/SESubstanceBolus.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package pulse.cdm.patient.actions;

import pulse.cdm.bind.PatientActions.SubstanceBolusData;
import pulse.cdm.bind.PatientActions.SubstanceBolusData.eRoute;
import pulse.cdm.properties.SEScalarMassPerVolume;
import pulse.cdm.properties.SEScalarVolume;
import pulse.cdm.substance.SESubstance;

public class SESubstanceBolus extends SEPatientAction
{

  private static final long serialVersionUID = -2177347285157137770L;
  protected eRoute adminRoute;
  protected SEScalarMassPerVolume concentration;
  protected SEScalarVolume dose;
  protected SESubstance substance;
  
  public SESubstanceBolus(SESubstance substance)
  {
    this.adminRoute = null;
    this.dose = null;
    this.concentration = null;
    this.substance = substance;
  }
  
  public void copy(SESubstanceBolus other)
  {
    if(this==other)
      return;
    super.copy(other);
    adminRoute = other.adminRoute;
    substance = other.substance;
    
    if (other.dose != null)
      getDose().set(other.getDose());
    else if (dose != null)
      dose.invalidate();
    
    if (other.concentration != null)
      getConcentration().set(other.getConcentration());
    else if (concentration != null)
      concentration.invalidate();
  }
  
  @Override
  public void reset()
  {
    super.reset();
    adminRoute = null;
    if (dose != null)
      dose.invalidate();
    if (concentration != null)
      concentration.invalidate();
  }
  
  @Override
  public boolean isValid()
  {
    return hasAdministrationRoute() && hasDose() && hasConcentration() && hasSubstance();
  }
  
  public static void load(SubstanceBolusData src, SESubstanceBolus dst)
  {
    SEPatientAction.load(src.getPatientAction(), dst);
    if(src.hasDose())
      SEScalarVolume.load(src.getDose(),dst.getDose());
    if(src.hasConcentration())
      SEScalarMassPerVolume.load(src.getConcentration(),dst.getConcentration());
    if(src.getAdministrationRoute()!=eRoute.UNRECOGNIZED)
    	dst.adminRoute = src.getAdministrationRoute();
  }
  
  public static SubstanceBolusData unload(SESubstanceBolus src)
  {
    SubstanceBolusData.Builder dst = SubstanceBolusData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SESubstanceBolus src, SubstanceBolusData.Builder dst)
  {
    SEPatientAction.unload(src,dst.getPatientActionBuilder());
    if (src.hasDose())
      dst.setDose(SEScalarVolume.unload(src.dose));
    if (src.hasConcentration())
      dst.setConcentration(SEScalarMassPerVolume.unload(src.concentration));
    if (src.hasAdministrationRoute())
      dst.setAdministrationRoute(src.adminRoute);
    if (src.hasSubstance())
      dst.setSubstance(src.substance.getName());
  }

  public eRoute getAdministrationRoute()
  {
    return adminRoute;
  }
  public void setAdministrationRoute(eRoute adminRoute)
  {
    this.adminRoute = adminRoute;
  }
  public boolean hasAdministrationRoute()
  {
    return adminRoute == null ? false : true;
  }
  
  public boolean hasConcentration()
  {
    return concentration == null ? false : concentration.isValid();
  }
  public SEScalarMassPerVolume getConcentration()
  {
    if (concentration == null)
      concentration = new SEScalarMassPerVolume();
    return concentration;
  }
  
  public boolean hasDose()
  {
    return dose == null ? false : dose.isValid();
  }
  
  public SEScalarVolume getDose()
  {
    if (dose == null)
      dose = new SEScalarVolume();
    return dose;
  }
  
  public SESubstance getSubstance()
  {
    return substance;
  }
  public boolean hasSubstance() { return substance != null; }
  
  @Override
  public String toString()
  {
    if (dose != null || concentration != null)
      return "Substance Bolus" 
          + "\n\tDose: " + getDose()
          + "\n\tConcentration: " + getConcentration()
          + "\n\tSubstance: " + getSubstance().getName() 
          + "\n\tAdministration Route: " + getAdministrationRoute();
    else
      return "Action not specified properly";
  }
}

File: src/csharp/pulse/cdm/patient/actions/SESubstanceBolus.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public enum eSubstanceAdministration_Route : int
  {
    Intravenous = 0,
    Epidural,
    Intraosseous,
    Intraarterial,
    Intracardiac,
    Intracerebral,
    Intracerebroventricular,
    Intradermal,
    Intramuscular,
    Subcutaneous
  }

  public class SESubstanceBolus : SEPatientAction
  {
    protected eSubstanceAdministration_Route admin_route;
    protected SEScalarTime admin_duration;
    protected SEScalarMassPerVolume concentration;
    protected SEScalarVolume dose;
    protected string substance;

    public SESubstanceBolus()
    {
      this.admin_route = eSubstanceAdministration_Route.Intravenous;
      this.admin_duration = null;
      this.dose = null;
      this.concentration = null;
      this.substance = null;
    }

    public override void Clear()
    {
      base.Clear();
      substance = null;
      admin_route = eSubstanceAdministration_Route.Intravenous;
      if (admin_duration != null)
        admin_duration.Invalidate();
      if (dose != null)
        dose.Invalidate();
      if (concentration != null)
        concentration.Invalidate();
    }

    public override bool IsValid()
    {
      return HasDose() && HasConcentration() && HasSubstance();
    }

    public eSubstanceAdministration_Route GetAdminRoute()
    {
      return admin_route;
    }
    public void SetAdminRoute(eSubstanceAdministration_Route adminRoute)
    {
      this.admin_route = adminRoute;
    }

    public bool HasAdminDuration()
    {
      return admin_duration == null ? false : admin_duration.IsValid();
    }
    public SEScalarTime GetAdminDuration()
    {
      if (admin_duration == null)
        admin_duration = new SEScalarTime();
      return admin_duration;
    }

    public bool HasConcentration()
    {
      return concentration == null ? false : concentration.IsValid();
    }
    public SEScalarMassPerVolume GetConcentration()
    {
      if (concentration == null)
        concentration = new SEScalarMassPerVolume();
      return concentration;
    }

    public bool HasDose()
    {
      return dose == null ? false : dose.IsValid();
    }

    public SEScalarVolume GetDose()
    {
      if (dose == null)
        dose = new SEScalarVolume();
      return dose;
    }

    public string GetSubstance()
    {
      return substance;
    }
    public bool HasSubstance() { return !string.IsNullOrEmpty(substance); }
    public void SetSubstance(string s)
    {
      substance = s;
    }
  }
}
